src/components/TrapFocus/TrapFocus.tsx

```diff
@@ -15,81 +18,98 @@ export interface TrapFocusProps {
   children?: React.ReactNode;
 }
 
-interface State {
-  shouldFocusSelf: boolean | undefined;
-}
-
-export class TrapFocus extends React.PureComponent<TrapFocusProps, State> {
-  state: State = {
-    shouldFocusSelf: undefined,
-  };
+export function TrapFocus({trapping = true, children}: TrapFocusProps) {
+  const [shouldFocusSelf, setFocusSelf] = useState<boolean | undefined>(
+    undefined,
+  );
 
-  private focusTrapWrapper: HTMLElement;
+  const focusTrapWrapper = useRef<HTMLDivElement>(null);
 
-  componentDidMount() {
-    this.setState(this.handleTrappingChange());
-  }
-
-  handleTrappingChange() {
-    const {trapping = true} = this.props;
-
-    if (this.focusTrapWrapper.contains(document.activeElement)) {
-      return {shouldFocusSelf: false};
+  const handleTrappingChange = () => {
+    if (
+      focusTrapWrapper.current &&
+      focusTrapWrapper.current.contains(document.activeElement)
+    ) {
+      return false;
     }
+    return trapping;
+  };
 
-    return {shouldFocusSelf: trapping};
-  }
-
-  render() {
-    const {children} = this.props;
-
-    return (
-      <Focus disabled={this.shouldDisable} root={this.focusTrapWrapper}>
-        <div ref={this.setFocusTrapWrapper}>
-          <EventListener event="focusout" handler={this.handleBlur} />
-          {children}
-        </div>
-      </Focus>
-    );
-  }
-
-  private get shouldDisable() {
-    const {trapping = true} = this.props;
-    const {shouldFocusSelf} = this.state;
+  useComponentDidMount(() => setFocusSelf(handleTrappingChange()));
 
+  const shouldDisable = () => {
     if (shouldFocusSelf === undefined) {
       return true;
     }
 
     return shouldFocusSelf ? !trapping : !shouldFocusSelf;
-  }
+  };
 
-  private setFocusTrapWrapper = (node: HTMLDivElement) => {
-    this.focusTrapWrapper = node;
+  const handleKeyDownLastElement = (event: any) => {
+    if (event.keyCode === Key.Tab && focusTrapWrapper.current) {
+      !event.shiftKey && focusFirstFocusableNode(focusTrapWrapper.current);
+
+      document.removeEventListener('keydown', handleKeyDownLastElement);
+    }
+  };
+
+  const handleKeyDownFirstElement = (event: any) => {
+    if (event.keyCode === Key.Tab && focusTrapWrapper.current) {
+      event.shiftKey && focusFirstFocusableNode(focusTrapWrapper.current);
+
+      document.removeEventListener('keydown', handleKeyDownFirstElement);
+    }
   };
 
-  private handleBlur = (event: FocusEvent) => {
+  const handleBlur = (event: FocusEvent) => {
     const {relatedTarget} = event;
-    const {focusTrapWrapper} = this;
-    const {trapping = true} = this.props;
 
     if (trapping === false) {
       return;
     }
 
+    if (focusTrapWrapper.current) {
+      const firstElement = findFirstFocusableNode(
+        findFirstFocusableNode(
+          focusTrapWrapper.current as HTMLElement,
+        ) as HTMLElement,
+      );
```

Comment:
The following functions find different scopes:

```js
const component = <div><modal>*</modal></div>;
findFirstFocusableNode(component) = will find the modal
findLastFocusableNode(component) = will find the content (*)
```

This is why I am finding the first element by first finding the container, then the desired first element. Will definitely want a test for this.

Reply:
The reason why findFirstFocusableNode finds the modal is because we added a tabIndex=-1 for accessibility so this workaround will not work for all scenarios. I've tried moving the divs around and it looks like ti would work.

Reply:
Ah good catch 🙈